### DISCUSSION TITLE
What is a proper way to define functions using RPC?

### DISCUSSION THREAD

**Discussion by sigasigasiga** [neovim/neovim#26736] [Q&A]
I've wrote a remote plugin that provides a Lua function that can be called by a neovim user.
Here's how it is currently implemented:
1. I get the channel id using `nvim_get_api_info`
2. Then I execute `nvim_exec_lua` with this code:
```lua
MyFunction = function()
    local id = --the channel id that I got from the call to `nvim_get_api_info`
    return vim.rpcrequest(id, 'AnRpcRequestCommand')
end
```

The implementation in the code can be found [here](https://github.com/sigasigasiga/neovim-dark-monitor/blob/6b540f36f23ed6cb0fb43b7fa3d9819d8c333b9d/src/monitor/service/request_handler.cpp#L48-L61) if you're interested.

As you can see, this is not the most beautiful solution to function declaration problem. Is there a better way to accomplish this task?

**Comment by HiPhish** [comment]
I think `nvim_exec_lua` is the way to go. How would you define executable code from outside the process anyway? What would be the lexical scope of the function? Maybe there is some answer to these questions out there, but to my knowledge in Neovim the solution is to send Lua code over to Neovim and evaluate it there.
